### PR TITLE
[MERGE] crm : fix and ease PLS settings and flow

### PR DESCRIPTION
**Pull request by dbeguin**
This commit applies multiples things:
- Fix PLS settings that were not loaded correctly (was not using technical field names)
- Add data for PLS settings to check all optional fields and set the start date.
- Set date pls field required
- use null value instead of empty for phone and email state fields to lessen the DB storage
- Remove Flanker that crashes anyway while validating a simple email address.
- Handle stage_id = False
- Remove default value for lead probability -> To let Bayes makes the magic.
- Set default pls date to today - 8 days to include all demo data.
- Set won should give 100% predicted probability

See sub commits for more details.

Task ID:2044539

### CODE REVIEW

File: addons/crm/models/res_config_settings.py [hunk 1]
```diff
@@ -15,8 +15,9 @@ class ResConfigSettings(models.TransientModel):
     module_crm_iap_lead_website = fields.Boolean("Create Leads/Opportunities from your website's traffic")
     module_crm_iap_lead_enrich = fields.Boolean("Enrich your leads automatically with company data based on their email address.")
     lead_mining_in_pipeline = fields.Boolean("Create a lead mining request directly from the opportunity pipeline.", config_parameter='crm.lead_mining_in_pipeline')
-    predictive_lead_scoring_start_date = fields.Date(string='Lead Scoring Starting Date', compute="_compute_pls_start_date", inverse="_inverse_pls_start_date_str")
-    predictive_lead_scoring_start_date_str = fields.Char(string='Lead Scoring Starting Date in String', default=date.today().strftime('%Y-%m-%d'), config_parameter='crm.pls_start_date')
+    predictive_lead_scoring_start_date = fields.Date(string='Lead Scoring Starting Date', compute="_compute_pls_start_date", inverse="_inverse_pls_start_date_str",
```
Comment (tde-banana-odoo) [inline addons/crm/models/res_config_settings.py]: Compute should not be required ... ensure a value in your compute method instead.

File: addons/crm/models/crm_lead.py [hunk 2]
```diff
@@ -1426,9 +1427,12 @@ def _cron_update_automated_probabilities(self):
 
         # Recompute all the leads. Won : probability = 100 | Lost : probability = 0 or inactive
         # Here, inactive won't be returned anyway
+        # Get also all the lead without probability --> These are the new leads. Activate auto probability on them.
         pls_start_date = self._pls_get_safe_start_date()
         if pls_start_date:
-            pending_lead_domain = [('probability', '<', 100), ('probability', '>', 0), ('create_date', '>', pls_start_date)]
+            pending_lead_domain = [('probability', '<', 100), ('probability', '>', 0)]
```
Comment (tde-banana-odoo) [inline addons/crm/models/crm_lead.py]: You could directly have
```
['&', ('create_date', '>', pls_start_date), '|', ('probability', '=', False), '&', ('probability', '<', 100), ('probability', '>', 0)]
```

File: addons/crm/models/crm_lead.py [hunk 3]
```diff
@@ -483,16 +481,32 @@ def _fields_view_get(self, view_id=None, view_type='form', toolbar=False, submen
...
 
+    def _rebuild_pls_frequency_table_threshold(self):
+        """ Called by action_set_lost and action_set_won.
+         Will run the cron to update the frequency table only if the number of lead is above
+         a specified value (from config_parameter) for onboarding purpose.
+         Once the threshold is reached, the config_param is set to 0 to avoid re-run the cron
+         and, mainly, to avoid making useless search_count in the future."""
+        pls_threshold = int(self.env['ir.config_parameter'].sudo().get_param('crm.pls_rebuild_threshold'))
+        if pls_threshold:
+            lead_count = self.env['crm.lead'].search_count([])
```
Comment (tde-banana-odoo) [inline addons/crm/models/crm_lead.py]: Should be sudo-ed, otherwise it depends on access rights ^^ .

File: addons/crm/models/crm_lead.py [hunk 4]
```diff
@@ -1374,6 +1392,16 @@ def _pls_get_naive_bayes_probabilities(self, batch_mode=False):
         save_team_id = None
         p_won, p_lost = 1, 1
         for lead_id, lead_values in leads_values_dict.items():
+            # if stage_id is null, return 0 and bypass computation
+            lead_fields = [values[0] for values in list(lead_values.get('values', {}))]
```
Comment (tde-banana-odoo) [inline addons/crm/models/crm_lead.py]: list(dict) -> get it keys. Is it really what you want ? Seems weird.

If it is the case, just to list (dict.keys()), it is easier to understand

**Comment by tde-banana-odoo** [comment]
@robodoo r+ rebase-merge

**Comment by tde-banana-odoo** [comment]
@robodoo r+ rebase-merge

**Comment by KangOl** [comment]
@robodoo p=1

**Pull request closed without merging**